gulpfile.js

```diff
@@ -68,50 +72,98 @@ const {validator, validatorWebui} = require('./build-system/tasks/validator');
-gulp.task('build', build);
-gulp.task('bundle-size', bundleSize);
-gulp.task('caches-json', cachesJson);
-gulp.task('check-exact-versions', checkExactVersions);
-gulp.task('check-links', checkLinks);
-gulp.task('check-owners', checkOwners);
-gulp.task('check-types', checkTypes);
-gulp.task('clean', clean);
-gulp.task('codecov-upload', codecovUpload);
-gulp.task('compile-jison', compileJison);
-gulp.task('create-golden-css', createGoldenCss);
-gulp.task('css', css);
-gulp.task('csvify-size', csvifySize);
-gulp.task('default', defaultTask);
-gulp.task('dep-check', depCheck);
-gulp.task('dev-dashboard-tests', devDashboardTests);
-gulp.task('dist', dist);
-gulp.task('e2e', e2e);
-gulp.task('firebase', firebase);
-gulp.task('generate-vendor-jsons', generateVendorJsons);
-gulp.task('get-zindex', getZindex);
-gulp.task('integration', integration);
-gulp.task('lint', lint);
-gulp.task('make-extension', makeExtension);
-gulp.task('nailgun-start', nailgunStart);
-gulp.task('nailgun-stop', nailgunStop);
-gulp.task('performance', performance);
-gulp.task('pr-check', prCheck);
-gulp.task('prepend-global', prependGlobal);
-gulp.task('presubmit', presubmit);
-gulp.task('prettify', prettify);
-gulp.task('process-3p-github-pr', process3pGithubPr);
-gulp.task('process-github-issues', processGithubIssues);
-gulp.task('serve', serve);
-gulp.task('size', size);
-gulp.task('todos:find-closed', todosFindClosed);
-gulp.task('unit', unit);
-gulp.task('update-packages', updatePackages);
-gulp.task('validator', validator);
-gulp.task('validator-webui', validatorWebui);
-gulp.task('vendor-configs', vendorConfigs);
-gulp.task('visual-diff', visualDiff);
-gulp.task('watch', watch);
+/**
+ * Creates a gulp task using the given name and task function.
+ *
+ * @param {string} name
+ * @param {function} taskFunc
+ */
+function createTask(name, taskFunc) {
+  checkFlags(name, taskFunc);
+  gulp.task(name, taskFunc);
+}
+
+/**
+ * Checks if the flags passed in to a task are valid.
+ * @param {string} name
+ * @param {function} taskFunc
+ */
+function checkFlags(name, taskFunc) {
+  if (!argv._.includes(name)) {
+    return; // This isn't the task being run.
+  }
+  const validFlags = taskFunc.flags ? Object.keys(taskFunc.flags) : [];
+  const usedFlags = Object.keys(argv).slice(1); // Skip the '_' argument
+  const invalidFlags = [];
+  usedFlags.forEach(flag => {
+    if (!validFlags.includes(flag)) {
+      invalidFlags.push(`--${flag}`);
+    }
+  });
+  if (invalidFlags.length > 0) {
+    log(
+      red('ERROR:'),
+      'Found invalid flags for',
+      cyan(`gulp ${name}`) + ':',
+      cyan(invalidFlags.join(', '))
+    );
+    log('For detailed usage information, run', cyan('gulp help') + '.');
+    if (validFlags.length > 0) {
+      log('Valid flags for', cyan(`gulp ${name}`) + ':');
+      validFlags.forEach(key => {
+        console.log(cyan(`\t--${key}`) + `: ${taskFunc.flags[key]}`);
+      });
+    }
+    process.exit(1);
```

Comment:
Instead of this, what if we throw an error or set `process.exitCode=1`? This would allow caller to handle the error instead of a hard exit and allows pending async operations to complete (useful when you run tasks in parallel).

Reply:
This check runs at the module level during task definition, before any `gulp` task begins execution, so it's guaranteed that there are no pending async operations going on when the check fails.

Couple more notes:

- Setting `process.exitCode = 1` will not prevent the gulp task from running with invalid flags. (We want to error out without running the task.)
- I tried throwing an error, but it prints a mostly useless stack trace.

<img width="1274" alt="Screen Shot 2020-02-14 at 9 50 50 PM" src="https://user-images.githubusercontent.com/26553114/74582720-3ff18f80-4f8d-11ea-8b6b-d74ef1ee2eed.png">